### ISSUE TITLE
--newdb should fail if stellar-core is already running

### ISSUE THREAD

**Issue by jedmccaleb** [stellar/stellar-core#408]

**Comment by HongxuChen** [comment]
I guess I can have a try for this enhancement, but not exactly sure how to define *already running*. Presently I think there are at least 2 options.

* if we detect a process named "stellar-core", we emit an error message and terminate;
* since `--newdb` requires `--conf FILE`, we can detect whether there are conflicts between this running and the existing instance from the configuration files, e.g, find whether there are same `DATABASE`, `HTTP_PORT`, `PEER_PORT`, `TMP_DIR_PATH`, `BUCKET_DIR_PATH`, `LOG_FILE_PATH` and also writable HISTORY locations.

The former approach is a bit defensive (in practice should suffice?), but the detection code may vary for Windows and *nix. The latter is more fine-grained, but there may be several corner cases to check.

What do you think is better?

**Comment by HongxuChen** [comment]
I made a naive [commit](https://github.com/StellarCN/stellar-core/commit/d64e0e3f9f8e8c5f2316f5a343582e482172d1b2) using the 1st approach for Linux (tested on Ubuntu 14.10). If that's acceptable, I may add corresponding code for Windows and Darwin.

**Comment by MonsieurNicolas** [comment]
The problem with approach 1 is that it renders local testing impossible (can't run several instances with different configs) and still leaves open really dangerous scenarios that we've seen already (see below).
I would go with option 2 as it allows to extend it over time to include scenarios that are crossing machine boundaries.

The things that are very dangerous to share are:
 1. bucket_dir_path (local only)
 2. database (should not be shared between machines)
 3. history (writeable archives should not be shared between machines)
As you see, each probably requires code specific to the scenario: 1 is some sort of file locking (there is one already, maybe it's buggy), 2 is doing it in SQL.
3 is probably the trickiest as it requires doing it in a generic fashion using the "put/get" exposed for history (generic). I am not sure it's 100% solvable (lack of a "test and set" type of operation), but it can probably cover a large portion of "bad configs/bad deployments" scenarios that could lead to writing to the same history (maybe by using some sort of lease mechanism).

I don't think we care testing port numbers as the process will just not start up if it can't bind to the port.

**Comment by HongxuChen** [comment]
Thanks for your instructions. I will look into option 2 further and try to deal with these issues you mentioned step by step.

**Comment by HongxuChen** [comment]
@MonsieurNicolas is there any need that all paths are canonicalized to absolute paths? The concern is that otherwise two configuration files may contain same local file path despite the string literals are different.

```
├── buckets
├── conf1      # BUCKET_DIR_PATH="buckets"
└── settings
    └── conf2  # BUCKET_DIR_PATH="../buckets"
```

For example, there is a `stellar-core --conf conf1` running in the directory containing `conf1`. Now we change directory to the one directly containing `conf2` and start `stellar-core --conf conf2 --newdb`. This will still does dangerous things in `buckets`.

**Comment by HongxuChen** [comment]
Will stellar-core permit configure like `DATABASE=sqlite3://dbname=stellar shared_cache=true` ? `soci` seems to allow it.

**Comment by HongxuChen** [comment]
The folders/files that are supposed to be private are not only for `--newdb` command, and the actual issue is that we should add a basic sanitizater for the config before doing destructive things (e.g., reset the database). Right?

**Comment by MonsieurNicolas** [comment]
re: allowing options to the DATABASE string, yes - but it should not matter for this issue. Like I mentioned earlier, whatever is done for the DATABASE setting should be done at the SQL layer, so that it works when using other backends like postgresql.

As far as sanity checks before doing anything destructive. I think that this issue is tracking the specific problem of simply running the instance and potentially causing damage to an other instance (running it is "destructive" when it comes to the buckets folder and database as it modifies them).

Note: for buckets, I think that https://github.com/stellar/stellar-core/blob/master/src/bucket/BucketManagerImpl.cpp#L101-L134 is probably sufficient, unless you see a problem with it (yes, there is flock on linux but it won't work on windows).

For the database/history, I imagine that maybe using some sort of ID that uniquely identifies the instance would work (and binding the database/history to a specific ID). The nodeID maybe? The problem with NodeID is that it's randomly generated for non-validator nodes. One way to fix that might be to persist the NodeID on the filesystem (again, only used for the non-validator case)?

**Comment by HongxuChen** [comment]
@MonsieurNicolas Yes, I think the RAII way of dealing with file lock should be more portable (I found an odd case where `stellar-core.lock` still resides when all instances of stellar-core exit, but couldn't reproduce it).

For database, maybe adding some checking before database object construction (https://github.com/stellar/stellar-core/blob/master/src/main/ApplicationImpl.cpp#L81)? The checking is like, if there is a file lock (a file named "${db_name}.lock" in current working directory, created inside database constructor) containing another nodeID, forbid writing and terminate.

And for history, the file lock containing nodeID can be put inside each writable history storage location; but maybe troublesome for remote directories. The sanity check can be put somewhere around call to `checkSensibleConfig` (https://github.com/stellar/stellar-core/blob/master/src/main/main.cpp#L237-L245).

But to avoid many file locks, maybe using a global file lock (in current working directory) is better (the content can be of toml format or something like that).

**Comment by HongxuChen** [comment]
Just noticed that creating a global lock file in current working directory doesn't solve the problem. Maybe should add a fixed name file lock (e.g., `.stellar.json.lock`) in a well known directory (e.g., `/tmp` or `~/` in Linux/Darwin)?

**Comment by MonsieurNicolas** [comment]
Not sure if you're still thinking about this one. The global lock is not an acceptable solution: it's actually perfectly fine to run multiple instances on the same machine, what is not OK is to run two instances that share the same data folder.

**Comment by HongxuChen** [comment]
@MonsieurNicolas Sorry I was a bit busy these days and didn't pay attention to this.

I mean the file lock can be used to contain several information of the current stellar-core instance, including pid, database URL etc. 

When another instance attempts to run with some dangerous behavior, it *firstly* checks whether there are suspicious files (for example, these files are named as `~/.${pid}.stellar.json.lock`; or alternatively one single file like `~/.stellar.json.lock` that contains many running instances). If there is, additionally checking whether there are conflicts for the corresponding files/folders.

Basically the idea is to extract all necessary dangerous information of one instance separately and make other instances aware of it.

**Comment by MonsieurNicolas** [comment]
yeah this is exactly what we're already doing for the bucket folder but wrong it seems :( I opened https://github.com/stellar/stellar-core/issues/883 to redo the file locking code.

Locally, the only dangerous local locations are the buckets folder and the temp folder (that we are not locking at the moment, but we should).

Using the lock file to track use of SQL database or history archives does not work as we want to avoid sharing them across machines.

I'm leaving this issue open to track the other resources that we do not want to share (SQL and history).

**Comment by HongxuChen** [comment]
@MonsieurNicolas  Okay, I think I get your concern now. 

So the focus should be how to deal with such a case: Machine `A` and `B` are running two instances of stellar-core and both store non-share info into `C`, and `C` is not running any instances of stellar-core; and we need to avoid conflicts between these two instances when the write targets are the same.

More concretely, stellar-core invokes native commands (`cp`/`aws`/`rsync`) to write these data, but currently there is no mechanism to keep these target locations safe since we cannot add a file lock *remotely* (and may have difficulties to determine disk locations of database?) and make other stellar-cores aware of it.

**Comment by MonsieurNicolas** [comment]
yes, you got it - that and the (remote) database; there might be a way to use some sort of lease mechanism.
The problem with time based leases is that in the event of a crash the process has to wait for the lease duration (sub optimal) or break the lease (which is dangerous).

**Comment by MonsieurNicolas** [comment]
pretty sure this was fixed by #1863 @marta-lokhova ?

**Comment by marta-lokhova** [comment]
@MonsieurNicolas that is correct. Just tried reproducing as a sanity check, and it fails, unable to acquire the lock (as expected).

**Comment by MonsieurNicolas** [comment]
cool thanks!
